### ISSUE TITLE
universal librairies for native plugins for Unity3D

### ISSUE THREAD

**Issue by roscopecoltran** [ruslo/polly#123]
Hi,

I could find a flag for universal/combined libraries for iOS but for MacOSX builds, how is it possible to lipo  flat libraries with i386/x86_64  and maybe rename the final output library filename for some platforms ?

On a wider spectrum, here is a quick overview of the goal:

## **What ?**
Add/Import in some way this script into polly for continuous integration of Unity3D native plugin:  https://github.com/mixpanel/mixpanel-unity/blob/master/native/build/build_all.py

eg. [OpenCV](https://github.com/hunter-packages/opencv) would need to generate the following libs:

**Output structure example** 

./Assets/Plugins/OpenCV/**Android/libs/armeabi-v7a**/**opencv-sharedlib.so**
./Assets/Plugins/OpenCV/**Android/libs/x86**/**opencv-sharedlib.so**
./Assets/Plugins/OpenCV/**iOS**/**opencv2.framework**
./Assets/Plugins/OpenCV/**iOS**/**libopencv.a** (combined armv7/arm64)
./Assets/Plugins/OpenCV/**opencv.bundle**
./Assets/Plugins/OpenCV/**WSA/PhoneSDK81/ARM**/***.dll** 
./Assets/Plugins/OpenCV/**WSA/PhoneSDK81/x86**/***.dll**
./Assets/Plugins/OpenCV/**WSA/SDK81/ARM**/***.dll**
./Assets/Plugins/OpenCV/**WSA/SDK81/x86**/***.dll**
./Assets/Plugins/OpenCV/**WSA/SDK81/x64**/***.dll**
./Assets/Plugins/OpenCV/**x86**/***.dll**
./Assets/Plugins/OpenCV/**x86_64**/***.dll**
**One dll per OpenCV module, eg. opencv_xfeatures2d310.dll, opencv_core310.dll ~27 files)**

## **Why ?**
In order to generate cross-platform libs, for iOS, MacOSX, Android, Windows and Linux libs to build native unity plugins, and to manage eveything with hunter+polly and use travis for CI. 

## **Notes**
For example, for creating a .bundle, you just have to rename .dylib output for MacOSX x86_63 and i386 to .bundle ?

Thanks for any thoughts, advices or maybe help on how to create this plugins for polly.

Have a great day !

Cheers,
Richard

**Comment by ruslo** [comment]
> I could find a flag for universal/combined libraries for iOS but for MacOSX builds, how is it possible to lipo flat libraries with i386/x86_64 and maybe rename the final output library filename for some platforms ?

Are we talking about Xcode? For iOS there is `CMAKE_XCODE_ATTRIBUTE_ONLY_ACTIVE_ARCH`, if there is similar property for MacOSX then building i386 + x86_64 is a trivial task, no need extra lipo step or anything.

> I saw a --framework flag in polly, is it related to the output of

`--framework` wrap dynamic/static library into `*.framework`. I don't know the details of OpenCV framework creation, but I think the version from Polly is simplier.

> How can we setup the arch to merge for combined libs on iOS and OSX ?

For iOS see `--ios-multiarch`/`--ios-combined`:
* http://polly.readthedocs.io/en/latest/toolchains/ios.html#ios-x-y

For OSX I'm not sure, see note above.

> How to specify if the output needs to be STATIC/SHARED per targeted platforms ?

https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html ?

> Add/Import in some way this script into polly for continuous integration of Unity3D native plugin: https://github.com/mixpanel/mixpanel-unity/blob/master/native/build/build_all.py

What do you think about managing this by Hunter instead?

**Comment by roscopecoltran** [comment]
Yes, with hunter it would be great too, I guess we need to create new function for mapping extension libs and install directories per platforms.

Any hints about how to manage the mapping of outputs per platforms (toolchains name are different than the platforms names expected by Unity3D), copy files in a separate install prefix folder (eg. PROJECT_UNITY3D_PLUGINS_PREFIX). I guess it would be in a **config.cmake** managed by HunterGate ?

**2 CMake driven examples buidling native unity3d plugins:**
1. https://github.com/ndob/yarrar (OpenCV, Catch, Eigen3, json11)
  - Covering iOS, Android, and desktop but not generate WP8, WP10 dlls.
2. https://github.com/reinterpretcat/utymap (Protobuf, Boost, ZLIB)
  - Not covering iOS, but super cool ! 

Last question, is it possible to simply trigger a request to get a copy of all/some hunter packages used as shared libs, and being copied in their respective ${PROJECT_UNITY3D_PLUGINS_PREFIX}/${PLATFORM_TOOLCHAIN_TARGET} ?
eg. Like Google.ProtocolBuffers.dll, for Google VR plugin (link: https://github.com/googlevr/gvr-unity-sdk/tree/master/Samples/CastleDefense/Assets/Plugins)

As a case study, I will hunterize [caffe-mobile](https://github.com/solrex/caffe-mobile) as it CMake based for iOS, Android and desktop.

Best,
Richard

**Comment by ruslo** [comment]
> I guess we need to create new function for mapping extension libs and install directories per platforms

I don't know all the Unity3D details but usually you only need to just install headers/libraries. Management of directories per each platform is a Hunter responsibility, see `toolchain-id`:
* https://docs.hunter.sh/en/latest/overview/customization/toolchain-id.html

> Last question, is it possible to simply trigger a request to get a copy of all/some hunter packages used as shared libs, and being copied in their respective ${PROJECT_UNITY3D_PLUGINS_PREFIX}/${PLATFORM_TOOLCHAIN_TARGET} ?

You can write such CMakeLists.txt code:
```cmake
hunter_add_package(foo)
find_package(foo CONFIG REQUIRED) # introduce foo::foo target
add_custom_command(
    ...
    COMMAND
    "${CMAKE_COMMAND}" -E copy
    "$<TARGET_FILE:foo::foo>"
    "/some/directory"
)
```

I'm not sure is it related to what you asking for but just for your information:
* https://github.com/ruslo/hunter/issues/274

> As a case study, I will hunterize caffe-mobile as it CMake based for iOS, Android and desktop.

For your info:
* https://github.com/ruslo/hunter/wiki/pkg.caffe
